docs/variance-defaulting.md

```diff
@@ -0,0 +1,375 @@
+  `Array` in type syntax produce an error.
+
+- Types for class literals (or more generally, any class with type templates).
+
+  All class singleton classes are generic classes (generic in the
+  `<AttachedClass>` type template that Sorbet implicitly declares). Without
+  defaulting rules, the type of an expression like `MyClass` would be
+  `T.class_of(MyClass)`. This is a bare generic, because type templates are just
+  [type members scoped to a singleton class].
+
+  No matter how hard we try, it would be impossible to ask users to replace
+  something like `MyClass.foo` with something like
+  `T.class_of(MyClass)[MyClass].foo` (even assuming this syntax were valid).
+  Indeed, most of our choices around type member defaulting rules were invented
+  to accommodate `<AttachedClass>`. But the rules were chosen from first
+  principles, not implementing something ad hoc for `<AttachedClass>`
+  specifically.
+
+- Instantiating a generic class and forgetting to provide type arguments.
+
+  This is _somewhat_ a repeat/special case of the previous, but worth mentioning
+  anyways. Consider code like this:
+
+  ```ruby
+  Array.new
+  Box.new
+  ```
+
+  Both of these are problematic. We'd prefer that users write something like
+  `Array[Integer].new` or `Box[String].new`. But because of how Sorbet builds
+  the CFG, both these look like
+
+  ```ruby
+  <tmp>$1 = Array
+  <tmp>$1.new()
+
+  <tmp>$2 = Box
+  <tmp>$2.new()
+  ```
+
+  Sorbet has to make a choice for the type of `<tmp>$1` and `<tmp>$2` when
+  they're assigned, which makes it hard for Sorbet to reject the call to `.new`
+  saying something like "you have to apply this class to a type when
+  instantiating it."
+
+  Unlike the previous case, `Box[Integer].new` is at least syntax we support,
+  because it's applying the arguments to the instance class. And even then, we
+  only support this syntax for the `.new` method, not for other singleton class
+  methods: `Box[Integer].make` isn't valid Sorbet syntax, because Sorbet
+  special-cases constructor methods.[^explicit-application]
+
+- Types for the `.class` method.
+
+  Consider code like this
+
+  ```ruby
+  class MyModel
+    extend T::Generic
+    MutatorType = type_template { {upper: MyMutator} }
+
+    def foo
+      klass = self.class
+    end
+  end
+  ```
+
+  What should the type of `klass` be in the snippet above? The type can't simply
+  be `T.class_of(MyModel)`, because that would be a bare generic. This is a very
+  similar problem to the class literal problem, where it would be prohibitively
+  annoying to ask users to rewrite code like `self.class`, and we don't even
+  have a suitable replacement if we wanted to enforce something.
+
+  Even more, there's a similar problem to the previous point:
+
+  ```ruby
+  my_box.class.new('')
+  ```
+
+  If `my_box` is `Box[Integer]`, there's no reason why `my_box.class.new` has to
+  create another `Box[Integer]`--making a `Box[String]` is fine!
```

Comment:
for what it's worth, I still don't know what behavior we should pick here. maybe there's some prior art in other languages we could copy here?

Reply:
i think the ideal solution here would be some way treat `new` on a generic class as if it had type `[T](x: T) -> Box[T]` like mentioned in the footnote. it's weird right now, because users specify the type of the constructor instead by specifying the type of the `initialize` method, where a `type_member` like `Elem` is in scope, and where the return type of the `initialize` method is always `void`.

I have chatted with dmitry about clever ways to make use of substitution to attempt to work around this but they all have varying levels of hackiness.

The raw notes from that conversation [are here](https://docs.google.com/document/d/1PDn3T4NeVsH0VMz4GdneqgUw6LcfenhrtD8bfOO8hek/edit)